packages/frontend/src/redux/actions/staking.js

```diff
@@ -533,3 +524,85 @@ export const handleUpdateCurrent = (accountId) => async (dispatch, getState) =>
     let currentAccount = selectStakingCurrentAccountbyAccountId(getState(), { accountId });
     dispatch(staking.updateCurrent({ currentAccount }));
 };
+
+export const getValidatorFarmData = (validator, accountId) => async (dispatch, getState) => {
+
+    if (validator?.version !== FARMING_VALIDATOR_VERSION || !accountId) return;
+
+    const poolSummary = await validator.contract.get_pool_summary();
+    const farms = await validator.contract.get_farms({ from_index: 0, limit: 300 });
+
+    const list = await Promise.all(farms.map(({ token_id, farm_id, active }) => {
+        try {
+            dispatch(fetchToken({ contractName: token_id }));
+            return validator.contract
+                .get_unclaimed_reward({ account_id: accountId, farm_id })
+                .catch(() => '0')
+                .then((balance) => ({ 
+                    token_id,
+                    balance,
+                    farm_id,
+                    active
+                }));
+        } catch (error) {
+            console.error(error);
+            return ({
+                token_id,
+                farm_id,
+                balance: 0,
+                active
+            });
+        }
+        
+    }));
+
+    const farmData = {
+        poolSummary: {...poolSummary},
+        farmRewards: list,
+    };
+    await dispatch(staking.setValidatorFarmData(validator.accountId, farmData));
+    return farmData;
+};
+
+export const claimFarmRewards = (validatorId, token_id) => async (dispatch, getState) => {
+    try {
+        const accountId = selectStakingMainAccountId(getState());
+        const currentAccountId = selectStakingCurrentAccountAccountId(getState());
+        const isLockup = currentAccountId !== accountId;
```

Comment:
This feels like a dangerous way to 'infer' if this is a lockup account. Don't we use a regex match to make this determination generally?